src/amo/components/HomepageShelves/index.js

```diff
@@ -0,0 +1,127 @@
+/* @flow */
+import * as React from 'react';
+import config from 'config';
+
+import LandingAddonsCard from 'amo/components/LandingAddonsCard';
+import LoadingText from 'amo/components/LoadingText';
+import {
+  INSTALL_SOURCE_FEATURED,
+  INSTALL_SOURCE_FEATURED_COLLECTION,
+  LANDING_PAGE_EXTENSION_COUNT,
+  LANDING_PAGE_THEME_COUNT,
+} from 'amo/constants';
+import translate from 'amo/i18n/translate';
+import type { ResultShelfType } from 'amo/reducers/home';
+import type { I18nType } from 'amo/types/i18n';
+
+import './styles.scss';
+
+type DefaultProps = {|
+  placeholderCount: number,
+|};
+
+type InternalProps = {|
+  i18n: I18nType,
+|};
+
+type Props = {|
+  ...DefaultProps,
+  ...InternalProps,
+  loading: boolean,
+  placeholderCount?: number,
+  shelves: Array<ResultShelfType>,
+|};
+
+class HomepageShelves extends React.Component<Props> {
+  static defaultProps: DefaultProps = {
+    placeholderCount: LANDING_PAGE_EXTENSION_COUNT,
+  };
+
+  render(): React.Node {
+    /* eslint-disable react/no-array-index-key */
+    const { i18n, loading, placeholderCount, shelves } = this.props;
+
+    const MOZILLA_USER_ID = config.get('mozillaUserId');
+
+    const homepageShelves: Array<React.Node> = shelves.map((shelf) => {
```

Comment:
The issue with this approach is that, even if `loading` is `true`, we are still doing some work to create `homepageShelves`. A different approach might be to have a variable called `shelvesContent` and have an `if` statement at the top setting `shelvesContent` to one thing is `loading` is `true` and to the other thing if `loading` is `false. That way you'll only build the shelf content you need in each case.

Reply:
I restructured the component and then tried to update it based on your comments, but I am not entirely sure if it's what you meant... please take a look.

Reply:
It's not exactly what I meant. I'll make a diff to show you what I am suggesting.